### DISCUSSION TITLE
virtio-fs use cases

### DISCUSSION THREAD

**Discussion by YanVugenfirer** [virtio-win/kvm-guest-drivers-windows#726] [Q&A]
Hello All!

If you are using virtio-fs with Windows VMs, we would like to hear from you about your use cases!

Please help us make a support of virtio-fs better.

Thanks a lot,
Yan.

**Comment by DocMAX** [comment]
Lots of apps in Windows do not work on shared network drives. I want a full physical emulated disk drive where i can install my apps on. That's MY use case.

**Reply by idarlund** [reply]
This is exactly the same reason I need it :)

**Reply by DocMAX** [reply]
Glad i'm not the only one. Please note, that there are some "virtual drive" solutions out there, but none has a TRUE virtual PHYSICAL disk drive. Without such a (device) driver Windows is still able to detect "fake" drives. Therefore, unfortunately an extra driver must be developed :-(.

**Reply by fabceolin** [reply]
Same reason

**Reply by fabceolin** [reply]
I am using this project https://github.com/fabceolin/windev/pkgs/container/windev to run windows on docker and I am wondering to mount a docker volume as a physical emulated disk drive.

**Reply by Marietto2008** [reply]
It would be nice to use a freebsd jail (equivalent of docker for linux) + bhyve (equivalent of qemu + kvm ) to virtualize windows on the jail and I am wondering if I can mount the jail as a physical emulated disk drive.

**Reply by HikariKnight** [reply]
Pretty much the same use case here too, store large files, lots of small files, compile or run programs from a shared filesystem directory that windows believes to be a physical disk. 

This also means that the filesystem needs the ability to optionally be case insensitive because so many developers of windows programs seem to use what i call MonkeyCASE typing with their filepaths in their code where they might look for file.dll but they type FILE.dll or File.DLL in their code, which then causes issues in these use cases 😂.

Currently this is achieved with iscsi but that makes the same device unable to be used in the host system while the guest VM is running, if i need to share with both systems at the same time i will have to setup a samba share which will have abysmal performance in comparison to virtio-fs.

**Reply by DocMAX** [reply]
exactly. that is the long explained version. the magic would make the yet non existent virtual device driver for a emulated physical disk.

**Reply by JustSomeGeek** [reply]
With some stuff you can install locally, move the program folder to a network drive, then edit the path in the registry. I've had success with a few things that wouldn't install to a network drive, but would happily run from one.

**Comment by ghomem** [comment]
Running windows server VMs on top of LXD (which uses KVM) in order to centralize legacy Window applications acessible via Terminal Services. The scenario is not only compatibility but performance too because some old applications are heavy and often Terminal Services and SQL server are mixed in the same VM.

Ideally performance would be comparable to native. Realistically, performance would be on par with the performance of Linux VMs (which is pretty good). See also #721 .

**Comment by agrisvv** [comment]
I run minimal linux workstation desktop with KVM. On top of it i have isolated environments VMs ( example: administrative access VM, daily apps and internet browsing VM, one for windows apps and tests.). Some of them linux, some windows. 
main usage: exchange of ISO files or some large archives, because upload and download somewhere vai internet are slow.

**Comment by tkolo** [comment]
I'm a .NET developer and sometimes need to launch Visual Studio for certain legacy projects. For that I have single Windows VM with a shared virtio drive to share the projects I need.

**Comment by GrandtheUK** [comment]
My main usage of virtio-fs would be file sharing and program/game sharing into the vm as i needed. however currently some executables do not run without killing the windows service while others work perfectly fine (for more see issue [#724](https://github.com/virtio-win/kvm-guest-drivers-windows/issues/724)).

**Comment by Marietto2008** [comment]
Hello.

I've virtualized Windows 11 with bhyve on FreeBSD 13R and It seems to me that the virtio-9p and or the virtio-fs driver does not work. I don't see the folder "mnt" mounted on the root folder of Windows 11. Below you can see which parameters I've used :

bhyve -S -c 4 -m 8G -w -H \
-s 0,hostbridge \
-s 1,ahci-cd,/home/marietto/Downloads/virtio/virtio-win-0.1.208.iso \
-s 2,ahci-hd,/mnt/da0p1/Backups/OS/bhyve/Windows/win11.img \
-s 3,fbuf,tcp=0.0.0.0:5900,w=1440,h=900,wait \
-s 8,virtio-net,tap0 \
----> -s 9,virtio-9p,sharename=/mnt \
-s 30,xhci,tablet \
-s 31,lpc \
-l bootrom,/usr/local/share/uefi-firmware/BHYVE_UEFI.fd \

It does not work because there are no virtio-9p drivers for Windows and bhyve does not support virtio-fs.

**Reply by idarlund** [reply]
This is not a thread for support for the virtio-fs, please check out the mailinglist or issue tracker.

**Reply by Marietto2008** [reply]
it makes no sense that I do that,since bhyve does not support virtio-fs. What's more interesting instead is to know why there are no virtio-9p drivers for Windows,because bhyve supports virtio-9p.

**Reply by YanVugenfirer** [reply]
Hi,

virtio-9p considered inferior to virtio-fs. Currently we don't have plans to develop a driver for it. If someone wants to do it, they can by using virtio-fs driver as an example.
And actually it make sense for bhyve to include support for more recent virtio devices.

**Reply by Marietto2008** [reply]
unfortunately we can't say what it makes sense for the bhyve developers to support.

**Comment by hiatsu0** [comment]
Been planning on attempting to configure some solution to allow Windows guest VM as well as some Linux containers to see a common shared folder on the host (Proxmox 7) and to transfer large media files to/from there. Ideally the folder would appear as a standard (or removable) drive in Windows and be available right at boot (as some of my Windows services will attempt to access it and start writing to in right after boot).

**Reply by Marietto2008** [reply]
the closer solution to your idea could be a tool called nitroshare. It should be installed in Windows guest and in FreeBSD host. Unfortunately there is no gui for the FreeBSD version. So it is not user friendly. I'm not sure if it works if I want to run docker for Windows for creating a tunnel and map it on a fixed disk when Windows starts. But Im sure that the linux version of Docker will not work on Windows using the WSL2 system because bhyve does not support the nesting virtualization. Another solution could be to install sshfs on Windows and on FreeBSD.

**Comment by Marietto2008** [comment]
It would be nice to use a freebsd jail (equivalent of docker for linux) +
bhyve (equivalent of qemu + kvm ) to virtualize windows on the jail and I
am wondering if I can mount the jail as a physical emulated disk drive.

Il mar 1 feb 2022, 01:22 Fabrício Ceolin ***@***.***> ha
scritto:

**Comment by daiaji** [comment]
Just for exchanging files between Linux host and Win guest, nothing special.

**Comment by YanVugenfirer** [comment]
Hello All,

Thank you very much for your feedback!

If someone is joining the discussion thread now, don't hesitate to leave your additional comments.

**Comment by XeCycle** [comment]
I work primarily on Linux but have a few C++/Rust projects that are either Windows-only or cross-Linux-and-Windows, so I need to compile and test on Windows.  My editor is Emacs running on Linux, so in an edit-compile-test cycle I edit on Linux, and compile-test on Windows; incremental recompilation requires a lot of "stat" ops so it is metadata-heavy load.

**Comment by KOLANICH** [comment]
Just wanna replace VirtualBox with QEMU because of better (para)virtual hardware options.

**Comment by jorins** [comment]
I prefer to run Linux in my root, but I use isolated Windows environments where needed. Professionally, that includes isolated administration machines, office application machines, etc. Recreationally, that includes gaming, music making, video editing. Things often have to go in and out of this. I may write software on my host machine and want to run it in a guest, download something that I want to install in a VM, or produce a song I want to share without logging into communication and management services on my creation VM.

I've looked at using SMB for this, which seems like the popular solution, but the setup is more complex and there are more security concerns so I have not felt confident setting up a SMB solution. From an architectural perspective, anything that requires network-based solutions for single-machine issues tickles me the wrong way. I'd also assume virtio-fs is able to achieve better transfer speed and is less prone to bugs stemming from having to implement protocols built for different purposes, though neither concerns me a lot (beyond #660 ;) )

**Comment by davispuh** [comment]
My main use case is to sometimes run windows applications/games inside KVM but sometimes using Wine. And to not duplicate data between Windows VM and host's it's very beneficial to have single storage. Especially as I'm using Linux MD RAID and BTRFS (subvolumes/snapshots very useful) and the only available solution is either Samba shares which I use most of time, or Virtio-FS.

Another issue is that I wish virto-fs appeared to support windows filesystem encryption (EFS), because some applications don't want to install to non-encrypted drives. I don't actually care if it's actually encrypted or not.

**Comment by dimavs** [comment]
It would be nice to do everything git/vim on linux host and build project (Embarcadero, you know :-) ) in Windows guest. It won't work right now (case-sensitive filesystem and Embarcadero likes to reference FILE.OBJ files).

**Comment by jandr** [comment]
My host runs Linux, but I do need and use some Windows VMs. I need to share files between the host and the VMs in a quick and reliable manner. While SMB/CIFS or SSH (sshfs) could work in some basic scenarios, I need the "shared folders" to be available also when the VMs are connected to full VPNs, which would restrict local network connectivity (so anything relying on SMB or SSH would simply not work, as it would be routed via the VPN). 

However, when using virtio-fs in the Windows guest, unfortunately I loose some features and flexibility, such as being able to save the state of the running VM (which would provide similar functionality as "hibernting" the OS).

**Reply by ReneH-cnx** [reply]
This is my use-case too.  I also was trying to have two different VioFS mount points shared to the Windows guest, but for some reason the WinFSP-VioFS driver ends up only seeing one, and mapping that to a drive letter.

**Comment by ChandonPierre** [comment]
We run both Windows and Linux VMs on [Kubevirt](https://github.com/kubevirt/kubevirt). VirtIO-FS allows us to share `RWX` storage volumes between Linux VMs without the overhead of running a Samba server. 

Unfortunately, the performance in Windows at the moment is not so good, so we primarily use Samba instead here.

**Comment by niraami** [comment]
I'd like for this driver to be on-par with VirtualBox's shared folders. I haven't used them in a while, but I never had this many issues with them. This is probably because I want to have Windows as my second system, but that's all it should be, all of the files should be shared with my Linux host, be it downloads, Steam games, projects (code, videos, photos, etc) so I don't have to think about what's on which system.

As some have said before me, emulating a real system disk is a must, some applications don't work, or choose to not work in network drives (which I assume the current driver behaves as). See below:
![image](https://user-images.githubusercontent.com/24532624/158614960-9cb2ab0e-5259-4abf-85b7-65a8ea3c251e.png "Git for Windows installation exe")
![image](https://user-images.githubusercontent.com/24532624/158615157-cc167c79-942f-4598-b793-da3e8383f4f6.png "Trying to run a PlatformIO project")

And there are many more.. i ran into at least 15 separate error messages when setting up my VM the other day. Solution? Copy the target files to the VM itself.. which is fine if it's not a 10GB+ installation file, game, etc..

Another issue is the stability of `virtiofs.exe`. It's borderline unusable, even with the [`queue` workaround](https://github.com/virtio-win/kvm-guest-drivers-windows/issues/527#issuecomment-917298904), as it crashes when doing specific things, like trying to edit projects in DaVinci Resolve with files from the virtual disk, or using it for cache folders in Adobe programs (Photoshop, AfterEffects, etc).  
Sometimes I got Resolve to work, but it would be slower than editing off of my 1GbE NAS... even though the passed storage is an NVMe SSD.. so performance should also be prioritized, as [WinFS is definitely not going to be a bottleneck](https://github.com/winfsp/winfsp/blob/master/doc/WinFsp-Performance-Testing.asciidoc)

All in all, libvirt is one of the most powerful virtualization tools, but is currently held back by the lack of storage sharing with Windows guests (I guess because it has little to no datacenter usability).

**Comment by dblueman** [comment]
I'm a server BIOS developer. The only OSs on these top-tier servers use is Linux, but ironically the BIOS build process requires the Visual Studio toolchain (Windows), so I share the codebase via SMB. VirtIO-FS isn't an option due to no case insensitivity option though, so build times are way higher than could be :(

- Dan

**Comment by bobobo1618** [comment]
My computer is a dual-purpose server + workstation that runs Linux. When I need to use Windows software (e.g. games), I boot a VM with PCI passthrough but all my bulk data are still managed by Linux.

Before Virtio-FS I just used SMB to share from host to guest but the performance wasn't very good. My hope is that Virtio-FS will have substantially better performance.

**Comment by 0xC0ncord** [comment]
I play games on Windows 10 with a passed through GPU for performance. Some games and other applications do not work over network shares. I want to highlight a few limitations with VirtIO-FS I have identified with my situation:

1) The sandboxing of the `virtiofsd` process on the Linux host causes underlying mounts to not be visible in the guest. For example, I have am sharing `/mnt/shared` to the guest, and I also have `/mnt/music` and `/mnt/games` mounted. These directories are bind-mounted to `/mnt/shared/{music,games}` but the guest cannot see these because of the sandbox.

2) We can workaround # 1 by mounting the filesystems in `/mnt/shared/{music,games}` and then bind-mounting *those* to `/mnt/{music,games}`, but the disk inside the guest will have the capacity of the filesystem it lives on. So, if the host filesystem (`/`) has a size of 32GiB, the disk in the guest will also be 32GiB while the underlying mounts can support much larger capacities. For example, if we want to install a large game to `Z:\shared\games` which is on a 3TiB partition, Steam will complain as it only sees a 32GiB disk.

I want to note that both of the above can be worked around by being able to mount multiple disks in the guest, but this is not currently possible as only one virtual disk can be visible at a time.

**Comment by jmbezekiel** [comment]
I do a lot of A/V projects as a hobby -- lots of pictures, lots of video, lots of audio. 4K videos and 24/192 audio eats up a ton of space, and large shoots I'll easily bring home 100+GB of data from one event. I use Linux on my main file server. I need to back that data up to the cloud in case of a disaster at the house. Currently I'm using backblaze B2, but it's prohibitively expensive for the volume of data in play.

What I'm trying to do (and it *almost* works) is use virtiofs to expose the media directories to the Windows VM, and use the standard consumer desktop Backblaze client to back all of it up. Unfortunately files bigger than 4M fail with 'TEMPORARY_OTHER' failure-- the backblaze client is doing something that isn't reading the files correctly. 

Yes, I could rsync it all into a windows VM and back it up that way, but the goal is to avoid needing twice the disk...

**Comment by Ethorbit** [comment]
Some things will only run (or only run well) on Windows so I spin them up in a Windows virtual machine with GPU passthrough, but I still want to be able to edit the files on Linux **while the VM is running.** 

I have a development VM with Unreal Engine where I manage my Unreal Projects and a gaming one where I play my Steam games. 

My hope is that eventually Virtio-FS will offer better performance as currently it is too slow for me as virtio disks are blazing fast in comparison and SMB seems to consistently have higher speeds.

**Comment by rokups** [comment]
I do some development work with visual studio / cmake, running some games that have huge data files, working with big archives (rar files in multiple tens of gigabytes). Doing all that over SMB, although not everything is working transparently (cmake cache directory somehow does not work on SMB share). Would be great to not need SMB for all that. Thank you for all the hard work 🙏

**Comment by galaxy001** [comment]
I want to share steam game files between Linux host and Windows guest which has vGPU mdev.
Their `steamapps` folders will be on different path of one ZFS pool, with dedup enabled.
So that I can try both Linux and windows version of a game.

A general case is to share large data files between software of different platforms. 
Like 3D models and rendered CG. The pipeline may involve tools on both Linux and windows.

**Comment by QORTEC** [comment]
I would be interested in running Windows on a VirtioFS, I'm thinking that [Quibble](https://github.com/maharmstone/quibble) may be able support this.

Currently my Windows VM uses ZFS Zvol as a HDD, however I would to switch to VirtioFS.
The switch should able easier access/share my data, eliminating the risk of file system corruption if data is accessed/changed form the VM host and guest in parallel.
This would also allow greater flexibility in how data is stored/managed, since guest (Windows) doesn't need to understand the hosts filesystem, the user to select the correct filesystem and features that makes sense for their use case.

**Comment by DocMAX** [comment]
What's the status?

**Comment by Cameronsplaze** [comment]
I'm looking for a way to share a home directory between a windows and linux VM. I haven't tried it yet, but Virtio-fs seems like the fastest solution if both VM's are running on the same machine.

**Comment by Marietto2008** [comment]
yeah,quibble is interesting. I'm on FreeBSD and I use a Windows 10 bhyve virtual machine that I run with this script :

```
bhyve -S -c sockets=1,cores=2,threads=2 -m 4G -w -H -A \
-s 0,hostbridge \
-s 1,nvme,/dev/nvd0,bootindex=1 \
-s 3,ahci-hd,/dev/$vmdisk4 \
-s 7,virtio-net,tap5 \
-s 10,hda,play=/dev/dsp,rec=/dev/dsp \
this does not work ------> -s 11,virtio-9p,sharename=/ \
-s 29,fbuf,tcp=0.0.0.0:5905,w=1500,h=950 \
-s 30,xhci,tablet \
-s 31,lpc \
-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_CODE.fd \
vm0:5 < /dev/null & sleep 2 && vncviewer 0:5
```

unfortunately windows 10 does not support the virtio-9p client and bhyve does not support virtio-fs.

**Comment by JustSomeGeek** [comment]
> 
> If you are using virtio-fs with Windows VMs, we would like to hear from you about your use cases!
> 
I use a mix of VM's (Mostly Linux & Windows), for gaming, other stuff, testing, and just for the fun of trying various OS's.

I want to share multiple host directories. eg I have a Games_Fast and Games_Slow dirs on m.2 and sata drives. I'd also want to share subdirs of ~, an ISO dir, and an untrusted dir for testing and sandboxing.

The guide for multiple instances of virtiofs using commandline doesn't seem to be working for me, and would be tedious to carry out on every Windows guest.

Using Linux as a guest for gaming, and having 2 Steam libraries on different shared dirs works great though, thanks!

**Comment by TheGreatMcPain** [comment]
I'm using a Windows 10 VM for gaming (especially VR since SteamVR and ALVR on Linux still need some work).

I'd like to use virtio-fs to share my Steam libraries between my host and quest, so I don't have to keep two copies the same game on the same SSD.

I prefer using something that's already included with qemu/libvirt, and I think setting up Samba shares would be too much hassle.

**Comment by skywurz** [comment]
I'm using a Windows 10 VM with Backblaze windows client to backup files on unraid /mnt/user directory. I had migrated from a physical file share windows box to unraid so all of the files live on unraid but i did not want to have to reupload to a different backup provider that supported linux. With this solution i was able to just inherit my backup state and continue backing up.

**Comment by EDI-Systems** [comment]
Using viofs to mount zfs on my Windows VM so I get access to those huge vaults w/games, songs, sw installers, photos, etc. Definitely saves your time to set up NFS or something. Just hope maybe in the future we have more performance, currently only useful for performing installations or pulling archives. Steam gaming seems to be too much still for the I/O performance, but for other uses this is more than fine.

**Comment by vs49688** [comment]
I have two use cases:
1. Keep my VMs disk relatively small (e.g. downloads go directly on the virtio-fs drive), and it works beautifully for this case.
2. Share a large C/C++ source tree between Linux and Windows.
   This is proving problematic as CMake can't resolve paths when running on the virtio-fs drive (MM-mounted)

**Comment by YanVugenfirer** [comment]
> I have two use cases:
> 
> 1. Keep my VMs disk relatively small (e.g. downloads go directly on the virtio-fs drive), and it works beautifully for this case.
> 2. Share a large C/C++ source tree between Linux and Windows.
>    This is proving problematic as CMake can't resolve paths when running on the virtio-fs drive (MM-mounted)

@vs49688 Thanks for sharing!
Can you please open an [issue](https://github.com/virtio-win/kvm-guest-drivers-windows/issues) for the problems with CMake? Thanks!

**Reply by vs49688** [reply]
I will try. I'm having some difficulty creating an isolated, reproducible example.

**Comment by gerard-ryan-immersaview** [comment]
Share a native git repo between my Linux and Windows VMs and build/debug/run the application in the source.
I switch between these VMs quickly making cross-platform changes and don't want to go via a git server to sync the source between them.
In some situations where it does work already, it's also convenient to quickly get the relevant installer on the host PC for distribution.

**Comment by pauschuu** [comment]
I run a Windows 11 Guest on an Arch Linux Host via libvirt. The Windows Guest is a full desktop guest with GPU passthrough and lots of performance optimizations like isolating certain CPUs on the host to be only used by the guest and so on.

I was hoping to find a very fast (as in SSD performance 1.7 GB/s file transfer) solution to share very large files between host and guest.

At the moment I think that running a Samba server on the host is likely to be a much faster solution.

**Comment by neutr0nst4r** [comment]
Share game installations between a Windows and a Linux VM for trying them out with Wine etc.. Unfortunately some game launchers (looking at you, EA app) do not seem to play too well with virtiofs, attempted both mount types.
